### ISSUE TITLE
Loss is getting zero when training Titanet with Adam

### ISSUE THREAD

**Issue by GrigorKirakosyan** [NVIDIA/NeMo#4583]
Hi

I am trying to train TitaNet with angular loss. When the optimizer is SGD everything is fine (loss decreases and TopK accuracy  increases). Switching optimizer from SGD to Adam or AdamW results zero loss and zero accuracy after some steps.  Do you have any clue about zero-loss?
Btw, number of speakers in my case is equal to ~25000.

**Comment by nithinraok** [comment]
Did you change the learning rate? try with 0.001 as the learning rate and wd=0.0002.

**Comment by nithinraok** [comment]
Please share your config file if possible.

**Comment by GrigorKirakosyan** [comment]
`name: &name "TitaNet"
sample_rate: &sample_rate 16000

model:
  train_ds:
    manifest_filepath: ???
    sample_rate: 16000
    batch_size: 64
    num_workers: 11
    shuffle: True

  validation_ds:
    manifest_filepath: ???
    sample_rate: 16000
    batch_size: 64
    shuffle: False

  model_defaults:
    filters: 512
    repeat: 3
    dropout: 0.1
    separable: true
    se: true
    se_context_size: -1
    kernel_size_factor: 1.0

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    normalize: "per_feature"
    window_size: 0.025
    sample_rate: *sample_rate
    window_stride: 0.01
    window: "hann"
    features: &n_mels 80
    n_fft: 512
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0
    pad_value: 0.0

  encoder:
    _target_: nemo.collections.asr.modules.ConvASREncoder
    feat_in: *n_mels
    activation: relu
    conv_mask: true

    jasper:
      -   filters: ${model.model_defaults.filters}
          repeat: 1
          kernel: [3]
          stride: [1]
          dilation: [1]
          dropout: 0.0
          residual: false
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat:  ${model.model_defaults.repeat}
          kernel: [7]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat: ${model.model_defaults.repeat}
          kernel: [11]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat: ${model.model_defaults.repeat}
          kernel: [15]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: &enc_feat_out 3072
          repeat: 1
          kernel: [1]
          stride: [1]
          dilation: [1]
          dropout: 0.0
          residual: false
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

  decoder:
    _target_: nemo.collections.asr.modules.SpeakerDecoder
    feat_in: *enc_feat_out
    num_classes: 24543
    pool_mode: 'attention'
    emb_sizes: 192
    angular: True

  loss:
    scale: 30
    margin: 0.2

  optim:
    name: "AdamW"
    params:
      lr: 0.001
      weight_decay: 0

    # scheduler setup
    sched:
      name: CosineAnnealing
      params:
        warmup_ratio: 0.1
        min_lr: 0.0

trainer:
  gpus: [0]
  accumulate_grad_batches: 1
  max_epochs: 50
  strategy: 'ddp'
  num_nodes: 1
  sync_batchnorm: True
  precision: 32
  val_check_interval: 0.33

exp_manager:
  exp_dir: null
  name: *name
  create_tensorboard_logger: True
  create_checkpoint_callback: True`

**Comment by GrigorKirakosyan** [comment]
I have used lr = 0.001 without weight decay

**Comment by nithinraok** [comment]
I would start experimenting by increasing batch_size to 128 and lowering lr to 0.0001, since you have only one GPU. Though adamw doesn't show changes with weight_decay, I would still keep it non-zero during debugging process. If possible try using wandb logging and share logs here.

**Comment by GrigorKirakosyan** [comment]
I noticed that weight normalization in decoder doesn't work properly. Can you take a look at this line? It seems that weight normalization doesn't applied properly.
https://github.com/NVIDIA/NeMo/blob/6442e339a47d30a106d869d1ef29cc1294753b75/nemo/collections/asr/modules/conv_asr.py#L874

**Comment by nithinraok** [comment]
I was training with adam optimizer and loss seems fine to me. how many samples in your dataset?

**Comment by GrigorKirakosyan** [comment]
I have 3749074 samples in my dataset. At first I also don't have this problem, but after some steps loss gets zero value. I fixed the line related to weight normalization.
Now it is working properly. Pls review weight normalization part(actually it doesn't change weights).

**Comment by nithinraok** [comment]
Thanks. Can you send a PR, will review the changes.

**Comment by nithinraok** [comment]
Closing due to inactivity

**Comment by ukemamaster** [comment]
> I have 3749074 samples in my dataset. At first I also don't have this problem, but after some steps loss gets zero value. I fixed the line related to weight normalization. Now it is working properly. Pls review weight normalization part(actually it doesn't change weights).

@GrigorKirakosyan It looks like you have a huge dataset. How long it took to generate the manifest file (train.json) from files list (train.txt)? I have similar size data and it is taking more than 20 hours still running. i run:
 `python NeMo/scripts/speaker_tasks/filelist_to_manifest.py --filelist path/to/train.txt --id -2 --out path/to/manifest.json --split`

**Comment by nithinraok** [comment]
There is an external PR that fixes the normalization part: https://github.com/NVIDIA/NeMo/pull/12246. We are not training any speaker models these days. 
Please verify with you setup if you can replicate and fix with this PR.
